src/hb-buffer-serialize.cc

```diff
@@ -379,6 +487,131 @@ hb_buffer_serialize_glyphs (hb_buffer_t *buffer,
   }
 }
 
+/**
+ * hb_buffer_serialize_unicode:
+ * @buffer: an #hb_buffer_t buffer.
+ * @start: the first item in @buffer to serialize.
+ * @end: the last item in @buffer to serialize.
+ * @buf: (out) (array length=buf_size) (element-type uint8_t): output string to
+ *       write serialized buffer into.
+ * @buf_size: the size of @buf.
+ * @buf_consumed: (out) (allow-none): if not %NULL, will be set to the number of byes written into @buf.
+ * @format: the #hb_buffer_serialize_format_t to use for formatting the output.
+ *
+ * Serializes @buffer into a textual representation of its content,
+ * when the buffer contains Unicode codepoints (i.e., before shaping). This is
+ * useful for showing the contents of the buffer, for example during debugging.
+ * There are currently two supported serialization formats:
+ *
+ * ## text
+ * A human-readable, plain text format.
+ * The serialized codepoints will look something like:
+ *
+ * ```
+ * <U+0651|U+0628>
+ * ```
+ * - Glyphs are separated with `|`
+ * - Unicode codepoints are expressed as zero-padded four (or more)
+ *   digit hexadecimal numbers preceded by `U+`
+ *
+ * ## json
+ * A machine-readable, structured format.
+ * The serialized codepoints will be a list of Unicode codepoints as
+ * decimal integers. For example:
+ * ```
+ * [1617,1576]
+ * ```
+ *
+ * Return value:
+ * The number of serialized items.
+ *
+ * Since: 2.7.3
+ **/
+unsigned int
+hb_buffer_serialize_unicode (hb_buffer_t *buffer,
+                             unsigned int start,
+                             unsigned int end,
+                             char *buf,
+                             unsigned int buf_size,
+                             unsigned int *buf_consumed,
+                             hb_buffer_serialize_format_t format)
```

Comment:
This function does not currently use any flags, but I’m wondering if it should take flags anyway for parity with the glyphs function and in case flags were needed in the future? (it should be documented as unused in this case).

Reply:
We're already different from the glyphs function in that we don't take a font parameter either. I can't imagine what flags you would want. But now that I think about it, I wonder if when serializing Unicode buffers we also want to output cluster IDs as well as codepoints...

Reply:
I’d make it take a font as well, but that is just me.

Reply:
Yeah, I'm going to rework this, and think about how to show clusters.

Reply:
> Yeah, I'm going to rework this, and think about how to show clusters.

Showing clusters is great idea. Same flags can be used. But taking font to serialize Unicodes buffer I can't see how helps.